File: app/src/main/java/com/example/myapplication/weight/xformode/EraserView.java
package com.example.myapplication.weight.xformode;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;

import com.example.myapplication.R;

public class EraserView extends View {
    public EraserView(Context context) {
        super(context);
        init();
    }

    public EraserView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public EraserView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    Paint paintEraser;
    Bitmap mBitmapSrc, mBitmapDst;
    Path path;
    float mPreX,mPreY;
    public void init() {
        //TODO 关闭硬加速
        setLayerType(View.LAYER_TYPE_SOFTWARE, null);

        paintEraser = new Paint();
        paintEraser.setStyle(Paint.Style.STROKE);
        paintEraser.setColor(Color.parseColor("#FF00FF"));
        paintEraser.setStrokeWidth(50);

        mBitmapSrc = BitmapFactory.decodeResource(getResources(), R.mipmap.guaguaka);
        mBitmapDst = Bitmap.createBitmap(mBitmapSrc.getWidth(), mBitmapSrc.getHeight(), Bitmap.Config.ARGB_8888);

        path = new Path();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int layerId = canvas.saveLayer(0, 0, mBitmapSrc.getWidth(), mBitmapSrc.getHeight(), null, Canvas.ALL_SAVE_FLAG);

        Canvas c = new Canvas(mBitmapDst);
        c.drawPath(path, paintEraser);
        canvas.drawBitmap(mBitmapDst, 0, 0, paintEraser);
        paintEraser.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_OUT));
        canvas.drawBitmap(mBitmapSrc, 0, 0, paintEraser);
        paintEraser.setXfermode(null);
        canvas.restoreToCount(layerId);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()){
            case MotionEvent.ACTION_DOWN:
                path.moveTo(event.getX(),event.getY());
                mPreX = event.getX();
                mPreY = event.getY();
                return true;
            case MotionEvent.ACTION_MOVE:
                float endX = (mPreX+event.getX())/2;
                float endY = (mPreY+event.getY())/2;
                path.quadTo(mPreX,mPreY,endX,endY);
                mPreX = event.getX();
                mPreY =event.getY();
                break;
            case MotionEvent.ACTION_UP:
                break;
        }
        postInvalidate();
        return super.onTouchEvent(event);
    }
}

File: app/src/main/java/com/example/myapplication/gitflow/weight/xformade/EraserView.kt
package com.example.myapplication.gitflow.weight.xformade

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.example.myapplication.R

class EraserView : View {
    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    var mBitmapSrc: Bitmap
    var mBitmapText: Bitmap
    var myPaint: Paint

    var myPath: Path = Path()
    var mLastX = 0F
    var mLastY = 0F

    init {
        myPaint = Paint().apply {
            isAntiAlias = true
            color = Color.WHITE
            style = Paint.Style.FILL_AND_STROKE
            strokeWidth = 50F
        }
        mBitmapText = BitmapFactory.decodeResource(resources, R.mipmap.guaguaka_text1)
        mBitmapSrc = BitmapFactory.decodeResource(resources, R.mipmap.guaguaka)
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        canvas?.apply {
            drawBitmap(mBitmapText, 0F, 0F, myPaint)

            val saveLayer = saveLayer(0F, 0F, mBitmapSrc.width.toFloat(), mBitmapSrc.height.toFloat(), null, Canvas.ALL_SAVE_FLAG)
            drawPath(myPath, myPaint)
            myPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OUT)
            drawBitmap(mBitmapSrc, 0F, 0F, myPaint)
            myPaint.xfermode = null

            restoreToCount(saveLayer)

        }
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        event?.apply {
            when (action) {
                MotionEvent.ACTION_DOWN -> {
                    myPath.moveTo(event.x, event.y)
                }
                MotionEvent.ACTION_MOVE -> {
//                    myPath.quadTo(mLastX, mLastY, (mLastX + event.x) / 2, (event.y + mLastY) / 2)
                    myPath.quadTo(mLastX, mLastY, event.x, event.y)
                    myPath.lineTo(event.x, event.y)
                }
            }
            mLastX = event.x
            mLastY = event.y
            postInvalidate()
        }
        return true
    }
}